app/com/linkedin/drelephant/tunin/pso/pso_param_generation.py

```diff
@@ -0,0 +1,252 @@
+
+    for individual in pso.archive:
+      archive.append(individual.__dict__)
+    for individual in pso._previous_population:
+      prev_population.append(individual.__dict__)
+    for individual in population:
+      current_population.append(individual.__dict__)
+
+    data['archive'] = archive
+    data['prev_population'] = prev_population
+    data['current_population'] = current_population # todo: send 2 or 3?
+    data['rnd_state'] = rnd_state
+    data_dump = json.dumps(data)
+    return data_dump
+
+def json_to_individual_object(json_list):
+  individuals = []
+
+  for element in json_list:
+    individual = inspyred.ec.ec.Individual()
+    # if 'candidate' in element:
+    #   individual.candidate = element['candidate']
+    # elif '_candidate' in element:
+    individual.candidate = element['_candidate']
+    individual.maximize = element['maximize']
+    individual.fitness = element['fitness']
+    individual.birthday = element['birthdate']
+    individuals.append(individual)
+  return individuals
+
+
+
+def main(json_tuning_state, display=False):
+  tuning_state = json.loads(json_tuning_state)
+  prng = Random()
+  args = {}
+
+  if 'archive' not in tuning_state:
+    prng.seed(time.time())
+    pso = restartable_pso.restartable_pso(prng)
+    pso.observer = inspyred.ec.observers.default_observer
+    pso.terminator = inspyred.ec.terminators.evaluation_termination
+    pso.topology = inspyred.swarm.topologies.ring_topology
+    population = pso.evolve(generator=initial_pop_generator, evaluator=dummy_fitness_evaluator, pop_size=3,
+                          bounder=bounder,
+                          maximize=False, max_evaluations=3, **args)
+
+    tuning_state = save_tuning_state(pso, prng, population)
+    print tuning_state
+
+  else:
+    archive = json_to_individual_object(tuning_state['archive'])
+    prev_population = json_to_individual_object(tuning_state['prev_population'])
+    initial_population = json_to_individual_object(tuning_state['current_population'])
+
+    str_rnd_state = tuning_state['rnd_state']
+    json_rnd_state = json.loads(str_rnd_state)
+    json_rnd_state[1] = tuple(json_rnd_state[1])
+    rnd_state = tuple(json_rnd_state)
+
+    prng.setstate(rnd_state)
+
+    pso = restartable_pso.restartable_pso(prng, _archive=archive,
+                         _previous_population=prev_population)
+    pso.observer = inspyred.ec.observers.default_observer
+    pso.terminator = inspyred.ec.terminators.evaluation_termination
+    pso.topology = inspyred.swarm.topologies.ring_topology
+
+    population = pso.evolve(seeds=[cs.candidate for cs in initial_population], initial_fit = [cs.fitness for cs in initial_population],
+                          generator=None, evaluator=dummy_fitness_evaluator, pop_size=3,
+                          bounder=bounder,
+                          maximize=False, max_evaluations=6, **args)
+    tuning_state = save_tuning_state(pso, prng, population)
+    print tuning_state
```

Comment:
What are the plans to test the tuning state? Why is this written in python? What was the rationale?

Reply:
We are using Particle Swarm Optimization algorithm for tuning the jobs automatically. We couldn't find a reliable implementation of this algorithm in Java and hence are using PSO implementation from inspyred python package.